internal/cloud/cloud.go

```diff
@@ -0,0 +1,96 @@
+package cloud
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/go-playground/validator/v10"
+	"github.com/sethvargo/go-envconfig"
+	apiv1 "github.com/substratusai/substratus/api/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const CloudEnvVar = "CLOUD"
+
+type Cloud interface {
+	// Name of cloud.
+	Name() string
+
+	// AutoConfigure attempts to determine cloud configuration from metadata servers, etc.
+	AutoConfigure(context.Context) error
+
+	// ObjectBuildImageURL returns the URL of the container image that should be used
+	// when Substratus builds a given Object.
+	ObjectBuiltImageURL(Object) string
+
+	// ObjectArtifactURL returns the URL of the artifact that was stored for a given Object.
+	ObjectArtifactURL(Object) *BucketURL
+
+	// AssociateServiceAccount associates the given service account with a cloud
+	// identity (i.e. updates annotations).
+	AssociateServiceAccount(*corev1.ServiceAccount)
+
+	// MountBucket mutates the given Pod metadata and Pod spec in order to append
+	// volumes mounts for a bucket.
+	MountBucket(*metav1.ObjectMeta, *corev1.PodSpec, ArtifactObject, MountBucketConfig) error
+}
+
+func New(ctx context.Context) (Cloud, error) {
+	var c Cloud
+	// If CLOUD is set, then pull configuration from environment variables.
+	cloudName := os.Getenv(CloudEnvVar)
+
+	if cloudName == "" {
+		if metadata.OnGCE() {
+			cloudName = "gcp"
+		}
+	}
+
+	switch cloudName {
+	case GCPName:
+		c = &GCP{}
+	default:
+		if cloudName == "" {
+			return nil, fmt.Errorf("unable to determine cloud, if running remotely, specify %v environment variable", CloudEnvVar)
+		} else {
+			return nil, fmt.Errorf("unsupported cloud: %q", cloudName)
+		}
+	}
+
+	if err := envconfig.Process(ctx, c); err != nil {
+		return nil, fmt.Errorf("environment: %w", err)
+	}
+
+	if err := c.AutoConfigure(ctx); err != nil {
+		return nil, fmt.Errorf("autoconfigure: %w", err)
+	}
+
+	if err := validator.New().Struct(c); err != nil {
+		return nil, fmt.Errorf("validation: %w", err)
+	}
+
+	return c, nil
+}
+
+type BucketMount struct {
+	BucketSubdir  string
+	ContentSubdir string
+}
+
+type MountBucketConfig struct {
+	Container string        // Example: trainer, loader
+	Name      string        // Example: model, model-saved, data
+	Mounts    []BucketMount // Example: model, data, logs
+	ReadOnly  bool
```

Comment:
Does this type correspond to a single fs mount or all of them on the container? `Mounts` has me think it captures all of them but `ReadOnly` should probably only apply to a single mount point and not all right?

Reply:
nice catch. I think ReadOnly needs to move to BucketMount type. Otherwise how would you express this?

model(readonly=False), model-saved(readonly=True), data(readonly=True)

The code further down makes me believe this might be a bug indeed: https://github.com/substratusai/substratus/pull/109/files#diff-a1c985bf4c51fa451b98fc1431485adc3910b38bf897053cd4a881323083008fR66

Reply:
GCSFuse allows for specifying readonly on an entire volume. The reason why I have multiple `.volumes` being appended is: I wanted to respect the bucket url reported by a given object instead of assuming that all objects in a given cluster are stored in the same bucket. One nice side effect of this: a Dataset object could eventually be imported into a GKE cluster that references a dataset from S3.

Reply:
Should probably rename to `type ObjectMount struct`?

Reply:
Thanks for the explanation - makes good sense now!